### PR TITLE
[Feature] move cold data to object storage without losing any feature…

### PR DESCRIPTION
**Pull request by deardeng**
…(FE)

Co-authored-by:platonekosama@gmail.com

# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/task/NotifyUpdateStoragePolicyTask.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        this.policyName = name;
+        this.properties = properties;
+    }
+
+    public TGetStoragePolicy toThrift() {
+        TGetStoragePolicy ret = new TGetStoragePolicy();
+
+        ret.policy_name = policyName;
+        ret.cooldown_datetime = Long.parseLong(properties.get(StoragePolicy.COOLDOWN_DATETIME));
+        ret.cooldown_ttl = Long.parseLong(properties.get(StoragePolicy.COOLDOWN_TTL));
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/task/NotifyUpdateStoragePolicyTask.java:47]: Currently implemented, only supported S3。If there are other types of requirements in the future, can modify based on this.

File: fe/fe-core/src/main/java/org/apache/doris/task/NotifyUpdateStoragePolicyTask.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        ret.s3_storage_param = new TS3StorageParam();
+        ret.s3_storage_param.s3_max_conn = Integer.parseInt(properties.get(S3Resource.S3_MAX_CONNECTIONS));
+        ret.s3_storage_param.s3_request_timeout_ms = Integer.parseInt(properties.get(S3Resource.S3_REQUEST_TIMEOUT_MS));
+        ret.s3_storage_param.s3_conn_timeout_ms = Integer.parseInt(properties.get(S3Resource.S3_CONNECTION_TIMEOUT_MS));
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/task/NotifyUpdateStoragePolicyTask.java:51]: well，i fix it.

File: fe/fe-core/src/main/java/org/apache/doris/task/DropReplicaTask.java [hunk 3]
```diff
@@ -23,11 +23,14 @@
 public class DropReplicaTask extends AgentTask {
     private int schemaHash; // set -1L as unknown
     private long replicaId;
+    private boolean isDropTableOrPartition;
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/task/DropReplicaTask.java:26]: well, i add some annotation.

File: fe/fe-core/src/main/java/org/apache/doris/service/FrontendServiceImpl.java [hunk 4]
```diff
@@ -992,6 +999,57 @@ public TWaitingTxnStatusResult waitingTxnStatus(TWaitingTxnStatusRequest request
...
+                iter -> {
+                    // default policy not init.
+                    if (((StoragePolicy) iter).getStorageResource() == null) {
+                        return;
+                    }
+                    TGetStoragePolicy rEntry = new TGetStoragePolicy();
+                    rEntry.setPolicyName(iter.getPolicyName());
+                    //java 8 not support ifPresentOrElse
+                    final long[] ttlCoolDown = {-1};
+                    Optional.ofNullable(((StoragePolicy) iter).getCooldownTtl())
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/service/FrontendServiceImpl.java:1018]: ok，i modify it more gracefully

File: fe/fe-core/src/main/java/org/apache/doris/service/FrontendServiceImpl.java [hunk 5]
```diff
@@ -992,6 +999,57 @@ public TWaitingTxnStatusResult waitingTxnStatus(TWaitingTxnStatusRequest request
...
-        throw new TException("not implement");
+        LOG.debug("refresh storage policy request");
+        TGetStoragePolicyResult result = new TGetStoragePolicyResult();
+        TStatus status = new TStatus(TStatusCode.OK);
+        result.setStatus(status);
+
+        List<Policy> policyList = Catalog.getCurrentCatalog().getPolicyMgr().getPoliciesByType(PolicyTypeEnum.STORAGE);
+        policyList.forEach(
+                iter -> {
+                    // default policy not init.
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/service/FrontendServiceImpl.java:1010]: done

File: fe/fe-core/src/main/java/org/apache/doris/service/FrontendServiceImpl.java [hunk 6]
```diff
@@ -992,6 +999,57 @@ public TWaitingTxnStatusResult waitingTxnStatus(TWaitingTxnStatusRequest request
...
+
+                    final long[] secondTimestamp = {-1};
+                    Optional.ofNullable(((StoragePolicy) iter).getCooldownDatetime())
+                        .ifPresent(date -> secondTimestamp[0] = date.getTime() / 1000);
+                    rEntry.setCooldownDatetime(secondTimestamp[0]);
+
+                    Optional.ofNullable(((StoragePolicy) iter).getMd5Checksum()).ifPresent(rEntry::setMd5Checksum);
+
+                    TS3StorageParam s3Info = new TS3StorageParam();
+                    Optional.ofNullable(((StoragePolicy) iter).getStorageResource()).ifPresent(resource -> {
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/service/FrontendServiceImpl.java:1030]: Currently implemented, only supported S3。If there are other types of requirements in the future, can modify based on this.

File: fe/fe-core/src/main/java/org/apache/doris/policy/StoragePolicy.java [hunk 7]
```diff
@@ -244,4 +314,100 @@ private static long getMsByCooldownTtl(String cooldownTtl) throws AnalysisExcept
         }
         return cooldownTtlMs;
     }
+
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/policy/StoragePolicy.java:317]: done

File: fe/fe-core/src/main/java/org/apache/doris/policy/StoragePolicy.java [hunk 8]
```diff
@@ -149,26 +190,55 @@ public void init(final Map<String, String> props) throws AnalysisException {
...
+        }
+
+        Resource r = checkIsS3ResourceAndExist(this.storageResource);
+        if (!((S3Resource) r).policyAddToSet(super.getPolicyName()) && !ifNotExists) {
+            throw new AnalysisException("this policy has been added to s3 resource once, policy has been created.");
+        }
+        this.md5Checksum = calcPropertiesMd5();
+    }
+
+    private static Resource checkIsS3ResourceAndExist(final String storageResource) throws AnalysisException {
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/policy/StoragePolicy.java:206]: Currently implemented, only supported S3. If there are other types of requirements in the future, can modify based on this.

File: fe/fe-core/src/main/java/org/apache/doris/policy/StoragePolicy.java [hunk 9]
```diff
@@ -115,12 +148,17 @@ public StoragePolicy(final PolicyTypeEnum type, final String policyName) {
         super(type, policyName);
     }
 
+    public static boolean isInteger(String str) {
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/policy/StoragePolicy.java:151]: done

File: fe/fe-core/src/main/java/org/apache/doris/policy/StoragePolicy.java [hunk 10]
```diff
@@ -149,26 +190,55 @@ public void init(final Map<String, String> props) throws AnalysisException {
...
         }
+        return resource;
     }
 
     /**
      * Use for SHOW POLICY.
      **/
     public List<String> getShowInfo() throws AnalysisException {
-        String props = "";
+        final String[] props = {""};
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/policy/StoragePolicy.java:222]: java lambda syntax, in the lambda function, i will change props[0] as ref. 
no use props[1], props[2].

File: fe/fe-core/src/main/java/org/apache/doris/qe/DdlExecutor.java [hunk 11]
```diff
@@ -304,6 +305,8 @@ public static void execute(Catalog catalog, DdlStmt ddlStmt) throws Exception {
             catalog.getPolicyMgr().createPolicy((CreatePolicyStmt) ddlStmt);
         } else if (ddlStmt instanceof DropPolicyStmt) {
             catalog.getPolicyMgr().dropPolicy((DropPolicyStmt) ddlStmt);
+        } else if (ddlStmt instanceof AlterStoragePolicyStmt) {
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/qe/DdlExecutor.java:308]: ALTER STORAGE POLICY has_test_policy_to_alter PROPERTIES("cooldown_datetime" = "2023-06-08 00:00:00");
use this sql to alter storage policy, so AlterStoragePolicyStmt is better.

File: fe/fe-core/src/main/java/org/apache/doris/policy/StoragePolicy.java [hunk 12]
```diff
@@ -19,29 +19,58 @@
...
  * Save policy for storage migration.
  **/
 @Data
 public class StoragePolicy extends Policy {
+    public static boolean checkDefaultStoragePolicyValid(final String storagePolicyName,
+                                                         Optional<Policy> defaultPolicy) throws DdlException {
+        if (!defaultPolicy.isPresent()) {
+            return false;
+        }
+
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/policy/StoragePolicy.java:66]: current, defaultPolicy is just support s3.

File: fe/fe-core/src/main/java/org/apache/doris/policy/PolicyMgr.java [hunk 13]
```diff
@@ -362,4 +398,30 @@ public static PolicyMgr read(DataInput in) throws IOException {
         policyMgr.updateMergeTablePolicyMap();
         return policyMgr;
     }
+
+    public Optional<Policy> findStoragePolicy(final String storagePolicyName) {
```
[the remark this answers is not in the document]
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/policy/PolicyMgr.java:402]: just a find storage policy function,

**Review by pengxiangyu — APPROVED** [review]

**Review by morningman — APPROVED** [review]

**Pull request merged**